[3 - 4단계 방탈출 예약 대기] 비타(오연택) 미션 제출합니다.

taek2222 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
[프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/4378)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
프로덕션 코드 보다는 테스트 코드에 더 집중했습니다! 그래서 리팩토링하지 못한 부분이 존재해 위와 같은 점수를 선정했습니다. 🙇🏻

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
소중한 시간 내어 리뷰해 주셔서 진심으로 감사합니다. 😊

### 테스트 집중한 개선 ✅ 
`Step2` 미션에서는 부족하다고 **느낀 테스트 전략을 집중적으로 학습하고 적용**해보았습니다. 

특히, `@WebMvcTest`와 `@DataJpaTest`를 중심으로 **슬라이스 테스트를 작성해보았으며, `Controller`와 `Repository` 테스트**에 많은 고민을 담았습니다. 

해당 부분에 대해 집중적으로 피드백 주시면 감사하겠습니다! ☺️

---

### 패키지 구조 ✅
현재는 하나의 도메인 패키지 내에서 `Presentation`, `Application`, `Domain` 레이어를 함께 구성하고 있습니다.

- `DTO`나 `Controller` 클래스가 각각 1개로 구성된 경우에는 별도의 하위 패키지로 분리하지 않았고,
- 클래스가 2~3개 이상일 경우에만 세분화하여 분리했습니다.

이러한 구조가 가독성과 유지보수 측면에서 적절했는지, 혹은 개선 여지가 있을지 의견을 듣고 싶습니다. 👂🏻

---

### 설계 선택 이유 ✅
LMS에서 제시한 테이블 분리 설계를 확인하기 전에, 기존 구조를 유지하면서도 기능을 구현할 수 있는 방법이 떠올라 현재와 같은 방향으로 구현하게 되었습니다. 

기존을 유지한 이유는 실무에서는 기존 코드를 최대한 건드리지 않고 빠르게 기능을 추가하는 연습도 필요하다고 생각했습니다. 🤔

- **LMS 설계안 반영 vs 기존 방식 유지**
- **Reservation의 상속 구조 설계 (Reserved, Waiting 등)**

다양한 설계 대안을 고민했고, **트레이드오프를 고려해 변경을 최소화**한 설계를 선택했습니다.

물론, 아서의 피드백을 바탕으로 더 나은 설계 방식에도 유연하게 도전할 준비는 되어 있습니다. 😆

---

### 슬라이스 테스트 ✅
`@DataJpaTest`는 큰 도움이 되었지만, `@WebMvcTest`를 활용한 Controller 테스트는 그 필요성을 크게 느끼지 못했습니다.

1. **인증/인가 처리 검증**
2. **HTTP 상태 코드 확인**
3. **요청/응답 JSON 형태 검증**
4. **요청 파라미터 처리 검증**

이 외에도 아서가 생각하시는 `@WebMvcTest`의 활용 이유나 관점이 있다면 꼭 공유 받고 싶습니다. 🤔

---

### 코드 리뷰 표시 ✅
**패키지 이동으로 인해 Files Changed가 다소 많을** 수 있습니다.

편하게 리뷰하실 수 있도록 실제 미션 구현(3~4단계) 부분에는 따로 표시를 남겨 두겠습니다. ✅

---

부족한 점이 있다면 언제든지 배우고 성장하고 싶습니다. 🔥
편하게 코멘트 남겨주시면, 배움과 성장의 기회로 생각하겠습니다!
정말 자유롭게 리뷰해주셔도 괜찮습니다! ☺️

**다시 한번 소중한 시간 내어 리뷰해 주셔서 진심으로 감사드립니다. 🙇🏻**

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

taek2222 committed:
test: 불필요한 테스트 클래스 제거

taek2222 committed:
refactor: 불필요한 `@EnableJpaRepositories` 임포트 제거

taek2222 committed:
refactor: 불필요한 주석 제거 및 코드 간결화

taek2222 committed:
feat: 예약 대기 버튼 추가 및 관련 로직 구현

taek2222 committed:
feat: 예약 상태에 대기 추가 및 대기 예약 생성 기능 구현

taek2222 committed:
feat: 예약 삭제 로직 수정 및 사용자/관리자 구분 추가

taek2222 committed:
feat: 예약 대기 생성 및 취소 API 연동

taek2222 committed:
feat: 예약 대기순위 반환 기능 추가

taek2222 committed:
refactor: 불필요한 유효성 검사 어노테이션 제거

taek2222 committed:
feat: 데이터베이스 초기 데이터 수정 및 예약 데이터 추가

taek2222 committed:
refactor: 예약 날짜 및 시간 검증 로직 수정

- validateIsPast 메서드를 validateReservationDateTime으로 변경
- 예약 날짜와 시간을 함께 검증하도록 로직 개선

taek2222 committed:
chore: 불필요한 주석 제거

- 예약 대기 취소 API 호출에 대한 TODO 주석 제거

taek2222 committed:
feat: 예약 대기 관리 화면 및 기능 추가

- 예약 대기 관리 HTML 템플릿 추가 (waiting.html)
- 예약 대기 관리 JS 스크립트 추가 (waiting.js)
- AdminController에 예약 대기 관리 화면 endpoint 추가

taek2222 committed:
feat: 예약 대기 관리 기능 구현

- 예약 대기 목록 조회 API 및 응답 데이터 처리 로직 추가
- 예약 대기 승인/거절 버튼 활성화 및 동작 구현
- 예약 대기 삭제 API 호출 로직 완료
- HTML 테이블에 예약 대기 순번 컬럼 추가

taek2222 committed:
feat: 대기 중인 예약 목록 조회 기능 추가

- 대기 상태의 예약 데이터를 조회하는 쿼리 메서드 추가
- ReservationService에 대기 중인 예약 목록 조회 로직 추가
- AdminWaitingReservationResponse DTO 추가 및 매핑 처리

taek2222 committed:
refactor: 대기 예약 관리를 위한 ReservationWaitController 분리

- 기존 ReservationController에서 대기 예약 관련 API 제거
- ReservationWaitController를 새로 추가하여 대기 예약 API 관리

taek2222 committed:
feat: 예약 대기 승인 기능 추가

- Reservation 도메인에 예약 상태 변경 메서드(`acceptStatus`) 추가
- ReservationService에 예약 대기 승인 비즈니스 로직 추가
- ReservationWaitController에 예약 대기 승인 API 엔드포인트 추가
- 프론트엔드에서 예약 대기 승인 API 호출 로직 구현

taek2222 committed:
refactor: Reservation 객체 생성 방식 변경

- Reservation 생성자 직접 호출을 Reservation.createReserved 정적 팩토리 메서드로 변경

taek2222 committed:
test: WAITING 상태 예약 조회 테스트 추가

- 날짜, 시간, 테마가 일치하는 최근 WAITING 상태 예약을 조회하는 테스트 케이스 추가

taek2222 committed:
feat: 대기 상태의 첫 예약 조회 메서드 추가

- 새 메서드 findFirstWaitingReservation(LocalDate, ReservationTime, Theme) 추가

taek2222 committed:
feat: 예약 삭제 시 대기 상태 처리 로직 추가

- 예약 삭제 시 첫 번째 대기 상태 예약을 승인 처리하는 로직 추가
- 관리자 예약 삭제 메서드에서도 동일 로직 적용

taek2222 committed:
chore: 코드 들여쓰기 규칙 통일 및 예약 데이터 추가

- JavaScript 파일 전반에 걸쳐 들여쓰기 규칙을 2칸으로 통일
- data.sql에 예약 상태(reserved) 데이터 1건 추가

taek2222 committed:
refactor: 도메인 및 패키지 구조 리팩토링

- Theme 및 ReservationTime 관련 클래스/인터페이스를 각자의 도메인별 패키지로 이동
- Status enum을 ReservationStatus로 변경
- 불필요한 static 변수 제거 및 final 필드 추가
- 관련된 import 경로 수정

taek2222 committed:
refactor: Admin 관련 클래스 및 경로 리팩토링

- AdminReservation 관련 클래스와 요청 경로를 admin 패키지로 이동
- AdminReservationController의 endpoint를 "/admin"에서 "/admin/reservations"로 수정
- 관련 import문과 테스트 코드 경로를 변경하여 패키지 변경 사항 반영

taek2222 committed:
test: 테스트 클래스 제거

- AdminControllerTest, ThemeControllerTest, UserControllerTest 삭제
- 불필요하거나 사용되지 않는 테스트 코드 정리

taek2222 committed:
refactor: 관리자 테마 관리 기능 분리

- AdminThemeController 생성 및 관리자 테마 생성/삭제 API 구현
- ThemeRequest에 toTheme 메소드 추가
- ThemeController에서 관리자 테마 생성/삭제 기능 제거
- ThemeService 리팩토링으로 코드 중복 제거 및 유효성 검사 개선

taek2222 committed:
test: 컨트롤러 테스트 추가

- AdminController 관련 테스트 추가
- ThemeController 관련 테스트 추가
- UserController 관련 테스트 추가

taek2222 committed:
refactor: 관리자 전용 멤버 조회 API 분리

- AdminMemberController.java 생성 및 `/members` 엔드포인트 구현
- 기존 MemberController에서 관리자 전용 `getMembers` 메서드 제거
- MemberService 및 MemberRepository의 패키지 경로 리팩토링

taek2222 committed:
refactor: 관리자 예약 및 시간 관리 API 분리 및 패키지 리팩토링

- 관리자 전용 예약 및 시간 관련 API를 별도 컨트롤러(AdminReservationWaitController, AdminReservationTimeController)로 분리
- 관련된 클래스 및 요청/응답 DTO를 admin.reservation 패키지로 이동
- 기존 컨트롤러(ReservationWaitController, ReservationTimeController)에서 관리자 기능 제거
- ReservationRepository에 @Repository 어노테이션 추가

taek2222 committed:
refactor: 관리자 예약 관련 클래스 패키지 이동

- `AdminReservationController`, `AdminReservationWaitController` 등 컨트롤러를 `presentation` 패키지로 이동
- DTO 클래스(`AdminReservationRequest`, `AdminWaitingReservationResponse`)를 `presentation.dto` 패키지로 이동
- 변경된 패키지 경로에 따른 코드 전반 수정

taek2222 committed:
test: AdminControllerTest와 UserControllerTest 리팩토링 및 TestCookieFixture 추가

- RestAssured를 MockMvc로 대체하고 WebMvcTest 환경으로 변경
- TestCookieFixture를 추가하여 테스트에서 쿠키 생성 로직 분리

taek2222 committed:
fix: JWT 토큰 prefix 값 수정

taek2222 committed:
refactor: MemberArgumentResolver 생성자 인자 타입 수정

taek2222 committed:
refactor: ThemeControllerTest 리팩토링 및 MockMvc 기반 전환

taek2222 committed:
test: 로그인 컨트롤러 테스트를 MockMvc 기반으로 리팩토링

taek2222 committed:
test: 회원 컨트롤러 테스트 리팩토링

- 기존 MemberControllerTest를 AdminMemberControllerTest로 변경 및 모듈화
- 테스트 구조를 Nested Class로 리팩토링하여 가독성 및 유지보수성 개선
- MockMvc와 Mockito를 사용해 테스트 방식을 표준화

taek2222 committed:
test: 관리자 예약 테스트 작성

- SpringBootTest에서 WebMvcTest로 변경하여 테스트 성능 최적화
- MockMvc를 활용한 관리자 예약 생성 및 삭제 테스트 추가ㅂ

taek2222 committed:
test: 관리자의 대기 예약 승인 API 테스트 추가

taek2222 committed:
test: AdminThemeController 테스트 코드 추가

- 테마 생성 API 테스트 추가 (성공 케이스 포함)
- 테마 삭제 API 테스트 추가 (성공 및 실패 케이스 포함)
- MockMvc와 Mockito를 활용한 테스트 구현

taek2222 committed:
test: 테스트 코드 리팩토링 및 예약 시간 응답 포맷 변경

- 테스트 코드를 RestAssured 기반에서 MockMvc 기반으로 전환
- 예약 시간 응답 DTO에서 LocalTime 타입을 문자열 포맷(HH:mm)으로 변경
- DateTimeFormatter 추가로 시간 데이터 포맷 명시적으로 관리

taek2222 committed:
test: Location 헤더 검증 테스트 추가

taek2222 committed:
test: 대기 예약 생성 API 테스트 추가

- 대기 예약 생성 성공 케이스 테스트 작성
- 유효하지 않은 요청으로 인한 실패 케이스 테스트 추가
- 이미 예약된 시간으로 인한 실패 케이스 테스트 추가

taek2222 committed:
test: 예약 컨트롤러 테스트 코드 작성

taek2222 committed:
test: 테스트 코드 정리 및 개선

- 관리자 대기 예약 승인 테스트 클래스명 수정 및 불필요한 테스트 제거
- Token 처리 로직에서 불필요한 문자열 제거 수정
- 예약 관련 테스트에서 사용되지 않는 코드 및 테스트 제거

taek2222 committed:
fix: refactor Member ID 생성 방식 수정

- Member 엔티티의 ID 생성 전략을 SEQUENCE에서 IDENTITY로 변경
- 테스트 데이터 SQL에서 ID 컬럼 제거 및 SEQUENCE 초기화 코드 삭제

taek2222 committed:
test: ThemeRepository의 인기 테마 조회 테스트 추가

- 최근 7일간 예약 수 기준으로 인기 테마를 조회하는 테스트 추가
- 테스트 데이터로 예약, 멤버, 테마, 예약 시간 관리 로직 구현
- 8일 이전 예약 데이터가 집계에서 제외되는지 검증

taek2222 committed:
fix: JWT 토큰 prefix 수정

taek2222 committed:
test: MemberRepository 테스트 추가

- 이메일로 회원을 조회하는 테스트 findByEmailSuccess 작성
- 존재하지 않는 이메일 조회 시 빈 Optional 반환 테스트 findByNonExistentEmailReturnsEmpty 작성

taek2222 committed:
test: 예약 관련 조회 및 검증 테스트 추가

- 특정 날짜와 테마 ID로 예약 조회 테스트 추가
- 회원의 모든 예약 정보와 대기 순위 조회 테스트 추가
- 대기 상태 예약 정보와 순위 조회 테스트 추가
- 특정 예약 시간 ID에 예약 존재 여부 확인 테스트 추가
- 특정 날짜와 시간으로 예약 존재 여부 확인 테스트 추가
- 특정 테마 ID에 예약 존재 여부 확인 테스트 추가

taek2222 commented on src/main/java/roomescape/admin/reservation/presentation/AdminReservationController.java:
```diff
@@ -34,6 +36,15 @@ public ResponseEntity<ReservationResponse> createReservation(
                 .body(reservation);
     }
 
+    @Auth(Role.ADMIN)
```
### 3-4 단계 미션 구현 코드 🚀
- 어드민 예약 취소

taek2222 commented on src/main/java/roomescape/admin/reservation/presentation/AdminReservationWaitController.java:
### 3-4 단계 미션 구현 코드 🚀
- 어드민 웨이팅 관리 코드

taek2222 commented on src/main/java/roomescape/admin/reservation/presentation/dto/AdminWaitingReservationResponse.java:
### 3-4 단계 미션 구현 코드 🚀
- 웨이팅 Response DTO

taek2222 commented on src/main/java/roomescape/reservation/application/ReservationRepository.java:
### 3-4 단계 미션 구현 코드 🚀
- JPQL 사용한 비즈니스 로직 구성

이번에, JPQL과 SQL 쿼리문 구성하는 연습을 시도하고자 객체지향적 보다는 쿼리 작성으로 로직을 구성해보았습니다!

taek2222 commented on src/main/java/roomescape/reservation/domain/Reservation.java:
```diff
@@ -36,24 +38,33 @@ public class Reservation {
-    public Reservation(final Member member, final Theme theme, final LocalDate date,
-                       final ReservationTime reservationTime) {
-        this(null, member, theme, date, reservationTime);
+    public static Reservation createReserved(final Member member, final Theme theme, final LocalDate date,
```
### 3-4 단계 미션 구현 코드 🚀
- 정적 팩토리 메서드로 예약 상태 생성자 분리

taek2222 commented on src/main/java/roomescape/reservation/domain/ReservationWithRank.java:
### 3-4 단계 미션 구현 코드 🚀
- 웨이팅 값 객체

taek2222 commented on src/main/java/roomescape/reservation/presentation/controller/ReservationWaitController.java:
### 3-4 단계 미션 구현 코드 🚀
- 웨이팅 등록 컨트롤러

taek2222 commented on src/main/java/roomescape/view/AdminController.java:
### 3-4 단계 미션 구현 코드 🚀
- 웨이팅 페이지 이동

taek2222 commented on src/main/java/roomescape/reservation/application/ReservationService.java:
### 3-4 단계 미션 구현 코드 🚀
- 웨이팅 관련 비즈니스 코드

taek2222 committed:
refactor: 예약 대기 승인 응답 본문을 제거

- `acceptWaitingReservation` 메서드의 반환 타입을 `ReservationResponse`에서 `Void`로 변경
- 응답 본문에 포함된 불필요한 `jsonPath` 검증 로직 제거

Hyunta commented on src/main/java/roomescape/reservation/presentation/dto/UserReservationsResponse.java:
```diff
@@ -15,16 +17,24 @@ public class UserReservationsResponse {
-    public Long getReservationId() {
-        return reservationId;
+    private String setReservedStatus(ReservationWithRank reservationWithRank) {
+        if (reservationWithRank.reservation().getStatus().equals(ReservationStatus.RESERVED)) {
+            return reservationWithRank.reservation().getStatus().getStatus();
```
해당 부분을 개선해볼 수 있을까요?

taek2222 replied:
**아래와 같이 개선했습니다.** 

테스트 코드에 집중하다보니, 전체적으로 객체지향을 지향하지 못했네요.. 😢

**변경 코멘트:**
- [refactor: Reservation 상태 확인 메서드 추가 및 리팩토링](https://github.com/woowacourse/spring-roomescape-waiting/pull/265/commits/5c925948fbffdbda99d5a4d93a20c9ec0a377e9e)

Hyunta commented on src/main/java/roomescape/theme/application/ThemeService.java:
```diff
@@ -20,40 +19,35 @@ public ThemeService(ReservationRepository reservationRepository, ThemeRepository
+    }
+
+    private void validateIsDuplicated(final Long themeId) {
+        if (reservationRepository.existsByThemeId(themeId)) {
```
메서드명과 하는 기능이 맞지 않는 것 같습니다.

taek2222 replied:
아래와 같이 변경했습니다. 꼼꼼하게 봐주셔서 감사합니다. ☺️

**변경 코멘트:**
- [refactor: 테마 삭제 검증 메서드 네이밍 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/265/commits/56ed49adaa3a762901a576ed0bcb7d56e803615a)

Hyunta commented on src/main/java/roomescape/theme/presentation/ThemeController.java:
```diff
@@ -0,0 +1,39 @@
+        );
+    }
+
+    @Auth(Role.GUEST)
+    @GetMapping("/popular")
+    public ResponseEntity<List<ThemeResponse>> getPopularThemes(
+    ) {
+        return ResponseEntity.ok().body(
+                themeService.getPopularThemes()
+        );
+    }
```
GUEST를 왜 따로 분리하셨나요?

taek2222 replied:
저도 페어의 코드로 시작해 `GUEST`의 궁금증이 있었습니다. 🤔

페어와 얘기를 나눠본 결과 `@Auth(ADMIN)`, `@Auth(USER)` 존재할 때, 명시하지 않은 경우 깜빡해서 작성하지 않았는지? 로그인 사용자가 필요 없는지? 의 생각 조차 하지 않기 위해 명확성을 중요하게 생각해 선언했다고 의견을 들었고, 저 또한 새로운 관점이라 배움의 기회라 생각하고 이를 수용하기로 했습니다!

Hyunta commented on src/test/java/roomescape/theme/presentation/ThemeControllerTest.java:
```diff
@@ -0,0 +1,134 @@
+
+    @Nested
+    @DisplayName("전체 테마 목록 조회 API")
+    class GetThemes {
+        @Test
+        @DisplayName("테마 목록을 반환한다")
+        void returnThemeList() throws Exception {
+            // given
+            List<ThemeResponse> themeResponses = List.of(
+                    new ThemeResponse(new Theme(1L, "테마1", "설명1", "썸네일1")),
+                    new ThemeResponse(new Theme(2L, "테마2", "설명2", "썸네일2"))
+            );
+
+            doReturn(themeResponses).when(themeService)
+                    .getThemes();
+
+            // when & then
+            mockMvc.perform(get("/themes"))
+                    .andDo(print())
+                    .andExpect(status().isOk())
+                    .andExpect(jsonPath("$[0].id").value(1))
+                    .andExpect(jsonPath("$[0].name").value("테마1"))
+                    .andExpect(jsonPath("$[0].description").value("설명1"))
+                    .andExpect(jsonPath("$[0].thumbnail").value("썸네일1"))
+                    .andExpect(jsonPath("$[1].id").value(2))
+                    .andExpect(jsonPath("$[1].name").value("테마2"))
+                    .andExpect(jsonPath("$[1].description").value("설명2"))
+                    .andExpect(jsonPath("$[1].thumbnail").value("썸네일2"));
+        }
```
Controller 테스트를 작성하시면서 어떤 부분을 중점적으로 보셨을까요?

taek2222 replied:
테스트를 작성하면서 되돌아보면, 특정 부분에 중점을 두기보다는 **`mockMvc`를 활용한 `Controller` 슬라이스 테스트**에 집중했던 것 같습니다. 그래도 과정에서 다음과 같은 기준을 세우고 테스트를 구성했습니다. 🤔

- `get~~` : **반환되는 Response Json 데이터 검증**
- `post~~` : **요청하는 ID와 Request 값, 반환되는 Response Json 데이터 검증**
- `인증/인가` : **유저의 권한에 따라 페이지 접속 여부 및 View 반환 검증**
- `예외처리` : **ControllerAdvice 전역 에러 처리 검증**

이러한 기준을 바탕으로 테스트를 설계하고자 노력했던 것 같습니다.

Hyunta commented on src/test/java/roomescape/view/AdminControllerTest.java:
```diff
@@ -1,78 +1,153 @@
+        }
+    }
+
+    @Nested
+    @DisplayName("대기 관리 페이지 테스트")
+    class AdminWaitingPage {
+        @Test
+        @DisplayName("ADMIN 권한은 대기 관리 페이지를 성공적으로 반환한다.")
+        void adminCanAccessWaiting() throws Exception {
+            Cookie adminCookie = testCookieFixture.createCookieWith(Role.ADMIN);
+
+            mockMvc.perform(get(ADMIN_BASE_URL + "/waiting").cookie(adminCookie))
+                    .andExpect(status().isOk())
+                    .andExpect(view().name("/admin/waiting"));
+        }
+
+        @Test
+        @DisplayName("인증되지 않은 사용자는 대기 관리 페이지를 접근할 수 없다.")
+        void unauthorizedCannotAccessWaiting() throws Exception {
+            mockMvc.perform(get(ADMIN_BASE_URL + "/waiting"))
+                    .andExpect(status().isUnauthorized());
+        }
```
인가 테스트까지 잘 구성해주셨네요👍🏻

Hyunta commented on src/test/java/roomescape/reservation/application/repository/ReservationRepositoryTest.java:
```diff
@@ -76,12 +220,20 @@ private record TestData(
+        entityManager.persist(reservation);
+
+        return reservation;
+    }
```
Fixture를 별도 클래스로 분리하지 않고 내부에서 만들도록 구현하신 이유가 있을까요?

taek2222 replied:
사실 처음엔 Fixture 클래스로 분리하는 걸 고려하지 않았습니다. 테스트 코드에 또 다른 클래스에 대한 의존성을 추가하는 것이 다소 어색하게 느껴졌기 때문입니다. 🤔

하지만 아서의 피드백을 통해 시도해보았고, **여러 방면에서 재사용할 수 있어 유용함**을 느꼈습니다. 😯

다만 **`Fixture`가 변경되면 해당 클래스를 사용하는 테스트 코드 전반에 영향**을 줄 수 있다는 점에서, **재사용성 향상이라는 장점과 테스트의 안정성 저하 가능성이라는 단점 사이의 트레이드오프가 존재**한다고 생각하게 되었습니다.

**변경 코멘트:**
- [변경 커밋 범위](https://github.com/woowacourse/spring-roomescape-waiting/pull/265/files/56ed49adaa3a762901a576ed0bcb7d56e803615a..6bbfad6258369acaff7f6adebf823ba2e9f21c6d)

Hyunta replied:
맞아요 말씀해주신 것처럼 Fixture를 클래스로 분리해서 여러곳에서 사용해준다면 변경사항이 생겼을 때 여러 테스트에 영향을 미치게 됩니다. 그런데 반대로 생각해보면 Fixture에 변화가 생겼다는 것은 클래스에도 변화가 생겼다는 의미이고 테스트가 깨져서 오히려 관련된 로직을 찾을 수 있으니까 좋은 역할을 한 것으로 볼 수 있지 않을까요?

taek2222 replied:
네 말씀해주신 내용 처럼, 클래스에 변화가 있다면 그 클래스를 의존하는 경우 에러가 발생할 가능성이 크다고 생각합니다. 🤔

그 변동 사항을 한 곳에서 관리할 수 있어서 더욱 관리가 용이하다고 생각했습니다.

Hyunta commented on src/main/java/roomescape/member/application/MemberRepository.java:
```diff
@@ -1,4 +1,4 @@
-package roomescape.member.application.repository;
+package roomescape.member.application;
```
> 현재는 하나의 도메인 패키지 내에서 Presentation, Application, Domain 레이어를 함께 구성하고 있습니다.
DTO나 Controller 클래스가 각각 1개로 구성된 경우에는 별도의 하위 패키지로 분리하지 않았고,
클래스가 2~3개 이상일 경우에만 세분화하여 분리했습니다.
이러한 구조가 가독성과 유지보수 측면에서 적절했는지, 혹은 개선 여지가 있을지 의견을 듣고 싶습니다. 👂🏻

패키지 구조를 어떻게 정할지는 철저히 팀의 의견과 기존 컨벤션에 따라서 정해지는 경우가 대부분이기 때문에 정답은 없습니다.
저도 제 의견을 드리지만 절대 정답은 아니라는 점만 다시 알아주세요.

도메인별로 패키지를 구성하게 되면 당연 가장 큰 장점은 도메인 별로 레이어를 확인하고 관리할 수 있기 때문에, 비교적 하나의 디렉토리에서 관리하는 파일의 갯수가 적다는 장점이 있을 것 같습니다. 그런데 만약 모호한 Service가 생기면 어떡하죠? Member와 Reservation 사이에 뭔가 조율해야하는 역할을 하는 Service가 있다면 어디에 위치시킬 것인가요? 사람마다 의견이 다를 수 있겠죠 그럼 결국 Service 클래스를 찾기 위해서 패키지를 헤메는 과정이 조금 추가될 것 같습니다.

만약 비즈니스가 너무 커서 팀마다 도메인을 따로 관리를 해야한다던지 관리를 따로해야하는 규모가 됐다면 도메인 별 패키지 구성이 괜찮아보이지만, 그전에 작은 규모에서는 도메인을 너무 세세히 나누게 되면 코드를 유지보수하는데 귀찮은 결정들을 많이 해야 합니다. 그런 부분들을 줄이기 위해서 도메인별 패키지 구조를 가져가더라도 최대한 큰 묶음으로 묶어서 관리를 하려고 할 것 같아요. 그래서 개인적으로는 지금 규모였다면 더 적은 패키지로 유지하려고 했을 것 같습니다.

taek2222 replied:
정말 좋은 말씀 감사합니다. 아직 도메인별로 패키지를 분리하는 작업이 조금은 서툴고, 도전적으로 접근하고 있는 단계입니다. 말씀해주신 내용 덕분에 많은 도움이 되었습니다. 다시 한 번 감사드립니다! 😊

Hyunta commented on src/main/java/roomescape/reservation/application/ReservationRepository.java:
```diff
@@ -0,0 +1,76 @@
+            """)
+    List<ReservationWithRank> findReservationsWithRankOfWaitingStatus();
+
+    @Query("""
+                 SELECT r FROM Reservation r
+                  WHERE r.date = :date
+                  AND r.reservationTime = :reservationTime
+                  AND r.theme = :theme
+                  AND r.status = 'WAITING'
+                  ORDER BY r.id ASC
+                  LIMIT 1
+            """)
+    Reservation findFirstWaitingReservation(LocalDate date, ReservationTime reservationTime, Theme theme);
```
JPQL에서 Limit이 작동하나요?

taek2222 replied:
네! 정상적으로 작동합니다. 📚

주변 크루의 이야기를 들었을 때, **_"JPQL는 Limit 지원은 안한다."_** 내용을 듣고, 궁금증이 들어서 실험해보았습니다.

근데, 정상으로 수행되는 것을 보고 자료를 탐색해본 결과 **하이버네이트 6.0 이상 부터 LIMIT 문법 지원**했으며, 내부 동작으로는 query.setMaxResults(..) 작동으로 된다는 것을 확인했습니다.

그래도.. 이론으로 접해서 혹시나 하는 마음에 실험해보았습니다!

<img src="https://github.com/user-attachments/assets/64cd8c9a-4f1c-4d17-a665-991cee322375" width="300"/>

Hyunta replied:
말씀해주신 것처럼 검색만 하고 끝내는 것이 아니라 직접 실험해보는 것은 중요하다고 생각해요.
좋은 경험을 공유해주셔서 감사합니다!

Hyunta reviewed (changes requested):
안녕하세요 비타,
이번 미션도 잘 구현해주셨네요.
몇가지 커멘트 남겨드렸는데 확인 부탁드립니다.

taek2222 committed:
refactor: Reservation 상태 확인 메서드 추가 및 리팩토링

- Reservation 클래스에 상태 확인 메서드(isReserved, getReservationStatus) 추가
- ReservationWithRank에 위 메서드 활용한 로직 추가
- UserReservationsResponse의 코드 리팩토링을 통해 가독성과 재사용성 향상

taek2222 committed:
refactor: 테마 삭제 검증 메서드 네이밍 수정

taek2222 committed:
test: 테스트 픽스처 클래스 추가 및 리팩터링

- Theme, Member, ReservationTime, Reservation 관련 테스트 픽스처 클래스 추가
- 기존 테스트 코드에 픽스처 활용하여 중복 제거 및 가독성 개선
- 인기 테마 조회 테스트 로직 수정 (데이터 생성 방식 개선, 테스트 대상 수 3 → 10)

taek2222 committed:
test: 테스트 fixture 클래스 및 메서드 리팩토링

- ReservationFixture 클래스 삭제
- 메서드 네이밍을 persist[Entity] 형태로 단순화
- 테스트 코드 내 불필요한 의존성 제거 및 가독성 개선

taek2222 committed:
test: 테스트 코드에서 MemberFixtures 클래스를 활용하도록 리팩토링

- Member 생성 및 영속화 메서드를 MemberFixtures로 통합
- 테스트 코드를 간결하게 리팩토링하여 중복 제거
- MemberFixtures에 기본값을 활용한 유틸리티 메서드 추가

taek2222 committed:
test: 예약 시간 테스트 추가

- 시작 시간으로 예약 시간 존재 여부를 확인하는 테스트 작성
- ReservationTimeFixtures에 기본 시작 시간을 사용할 수 있는 메서드 추가

taek2222 committed:
test: 테스트 코드에서 Fixture 활용

- ReservationTests 및 ThemeRepositoryTests에서 Fixture 유틸 방식으로 변경
- createAndSave 메서드 제거 및 Fixture 메서드 호출로 대체
- 테스트 코드의 가독성과 재사용성 개선

taek2222 committed:
test: ReservationFixtures 및 테스트 코드 개선

- ReservationFixtures에 다양한 메서드 오버로드 추가로 가독성과 재사용성 향상
- 테스트에서 flushAndClear 사용으로 데이터 일관성 보장
- Repository 메서드 이름을 findAllByDateAndThemeId로 수정하여 명확성 증대

taek2222 committed:
test: 토큰 파싱 로직 수정

taek2222 committed:
refactor: 예약 검증 및 로직 분리 리팩토링

- ReservationValidator 클래스 추가로 예약 검증 로직 분리
- ReservationService의 중복된 검증 및 로직을 ReservationValidator와 하위 서비스로 이동
- 불필요한 의존성 제거 및 관련 서비스와 리포지토리를 의존성 주입으로 대체
- MemberService, ThemeService, ReservationTimeService에서 조회 메서드 통합 및 메서드 네이밍 개선
- SignUpRequest 객체에 Member 변환 로직 추가 및 관련 로직 단순화

taek2222 commented:
- [refactor: 예약 검증 및 로직 분리 리팩토링](https://github.com/woowacourse/spring-roomescape-waiting/pull/265/commits/deb231b6836e784a0ab2e26482df589ac3b07024)

마지막 커밋에서는 비즈니스 로직을 최대한 메서드로 분리하고, 전체적인 흐름을 개선해보았습니다.

소중한 시간 내어 리뷰해주셔서 진심으로 감사드립니다. 🙇🏻

마지막 날까지 다양한 피드백을 통해 더 많이 배우고 성장하고 싶습니다. 감사합니다! ☺️

Hyunta reviewed (approved):
안녕하세요 비타,
피드백 반영도 빠르게 잘해주셨네요.
남겨주신 커멘트도 확인했습니다.
추가로 나눠보고 싶은 이야기가 있으면 DM 주세요.
이번 단계는 여기서 머지하도록 하겠습니다.

jeyongsong commented on src/main/java/roomescape/global/jwt/JwtTokenProvider.java:
```diff
@@ -13,21 +13,20 @@
     private static final String JWT_EXCEPTION_MESSAGE = "잘못된 로그인 시도입니다. 다시 시도해 주세요.";
+    private static final String PREFIX = "token=";
 
     @Value("${security.jwt.token.secret-key}")
     private String secretKey;
     @Value("${security.jwt.token.expire-length}")
     private long validityInMilliseconds;
```
JWT 관련 설정 값을 각각 `@Value`로 주입받아 사용하는 부분은,  
`@ConfigurationProperties`를 활용해 **설정 클래스로 캡슐화**하면 더 의미 있는 단위로 관리할 수 있어 보입니다.

데이터는 일반적으로 클래스(함수)보다 변경과 관리가 더 까다로운 특성이 있습니다.
하지만 설정 정보를 하나의 클래스(함수)로 캡슐화하면,
이 데이터를 변경하고 사용하는 모든 흐름을 하나의 통로에서 제어할 수 있게 되어
검증, 후처리, 테스트 주입, 리팩토링 등의 측면에서 훨씬 유연하고 안정적인 구조를 만들 수 있습니다.

관련 리팩토링 항목: 『리팩터링 2판』 6.6 _"변수 캡슐화하기 (Encapsulate Variable)"_

taek2222 replied:
**정말 새로운 점 알게 되어서 감사합니다! ☺️** 우선 두 개의 값 밖에 없지만, 적용을 해보았습니다.

하면서 느낀 점은 클래스 분리로 생성 비용이 증가하지만, 해당 환경 변수를 캡슐화해 행동을 위임할 수 있었던 점입니다.

그치만, 머피에게 궁금한 점이 생겼습니다. 값을 바인딩 하기 위해서는 **setter를 명시해야해 불안전한 객체가 된다는 점**이랑 아래 사진과 같이 사용하고 있음에도 불구하고, **'사용하지 않음' 에러가 뜬다는 점**입니다. 이를 개선할 수 있는 방법이 있을까요? 🤔

![스크린샷 2025-05-28 오전 9 59 35](https://github.com/user-attachments/assets/a20e2800-8d14-447f-a44b-538f6a6df5eb)

**변경 코멘트:**
- [refactor: JwtProperties 클래스를 추가하고 JwtTokenProvider에 이를 주입하도록 수정](https://github.com/taek2222/spring-roomescape-waiting/commit/42c52383722316bd497f1a797c3bc0237b66997e)
- [test: JwtProperties 클래스를 테스트 환경에 추가](https://github.com/taek2222/spring-roomescape-waiting/commit/44d4e9a81c31f578d5e288d6437aaedd21c1a8e7)
- [test: JwtProperties를 활용하여 TestCookieFixture 리팩토링](https://github.com/taek2222/spring-roomescape-waiting/commit/de38b8259c121fab6602d9bfb15ffc846267b395)

jeyongsong replied:
> 값을 바인딩 하기 위해서는 setter를 명시해야 해 불완전한 객체가 된다는 점

[@ConfigurationProperties로 final 필드 바인딩하기](https://medium.com/@nayoung238/configurationproperties%EB%A1%9C-final-%ED%95%84%EB%93%9C-%EB%B0%94%EC%9D%B8%EB%94%A9%ED%95%98%EA%B8%B0-63a4abbcbc4a) 이 글이 꽤 도움될 것 같아요!  
그리고 제 기억엔  생성자를 사용하는 방식으로 final 필드 바인딩이 가능하긴한데,  
이게 Gradle 환경에서만 되는 걸로 알고 있어요 (기억이 가물가물하지만요…)

그리고 `@Component`로 등록해서 사용하는 것도 괜찮지만, 
글에서 소개하는 다른 방법들도 참고해보셔도 좋을 것 같네요!

> 사용하고 있음에도 불구하고, '사용하지 않음' 에러가 뜬다는 점

이 부분은 저도 잘 모르겠네요… ㅠㅠ IDE 문제일 수도 있을 것 같아요!

jeyongsong commented on src/main/java/roomescape/reservation/application/ReservationService.java:
```diff
@@ -0,0 +1,155 @@
+
+    @Transactional
+    public ReservationResponse createReservation(final ReservationRequest request, final Long memberId) {
+        return createAndSaveReservationBy(
+                memberId,
+                request.getThemeId(),
+                request.getDate(),
+                request.getTimeId(),
+                false
+        );
```
`createAndSaveReservationBy` 메서드는 여러 개의 개별 매개변수를 받아 사용하고 있으며,  
여러 메서드에서 반복적으로 호출되고 있습니다.

이러한 구조는 메서드 호출 시 각 값의 의미를 파악하기 어렵고,  
매개변수의 순서가 바뀌거나 추가될 경우 유지보수가 까다로워질 수 있습니다.

현재 전달되는 매개변수들은 대부분 하나의 동작(예약 저장)을 위한 관련된 데이터들이므로,  
이를 별도의 매개변수 객체로 묶어 전달하면 더 의미 있는 단위로 관리할 수 있어 보입니다.

데이터를 뭉쳐서 구조화하면, **데이터 간의 관계가 명확해진다는 장점**이 있습니다.  
또한 함수가 해당 데이터 구조를 매개변수로 받게 되면  
**매개변수 수가 줄어들고**,  
**공통된 이름을 통해 일관성 있는 코드 작성이 가능**해집니다.

관련 리팩터링 항목: 『리팩터링 2판』 6.8 _"매개변수 객체 만들기 (Introduce Parameter Object)"_

taek2222 replied:
파라미터 인자가 여러 존재한다는 느낌이 들었을 때, 머피의 의견이 충분히 공감되었습니다. 그래서 바로 리팩토링을 실시해봤습니다. 🤔

결과적으로 볼 때, 인자가 줄었다는 장점이 있었지만, 객체의 행동도 없고, 단순히 데이터 전달 객체라는 느낌이 강하다고 느껴집니다. 과연 클래스를 분리해 관리할 정도로 비용이 맞을까라고 생각했습니다. 이에 대한 머피 의견이 궁금해요!

**변경 코멘트:**
- [refactor: Reservation 생성 로직 리팩터링](https://github.com/taek2222/spring-roomescape-waiting/commit/c64d5bba1f77e76949fcb13efec48f94bba531c4)

jeyongsong replied:
ㅎㅎ 이에 대해서는 저도 확신이 있진 않네요!  
캡슐화를 통해 얻을 수 있는 장점과 현재 구조 사이의 **트레이드오프**를 고민해볼 수 있는 부분인 것 같아요.

다만 지금 프로젝트의 **규모**나, **RESTful API 기반의 레이어드 아키텍처**로 구성되어 있는 점을 고려했을 때는  
저도 비타 처럼, 장점이 크게 부각되지는 않는다고 생각합니다 :)

jeyongsong commented on src/main/java/roomescape/reservation/application/ReservationService.java:
```diff
@@ -0,0 +1,155 @@
+                .orElseThrow(() -> new NoSuchElementException("예약 정보를 찾을 수 없습니다."));
+    }
+
+    private ReservationResponse createAndSaveReservationBy(
+            final Long memberId,
+            final Long themeId,
+            final LocalDate date,
+            final Long reservationTimeId,
+            final boolean isWaiting
+    ) {
+        Member member = memberService.getMemberById(memberId);
+        Theme theme = themeService.getThemeById(themeId);
+        ReservationTime reservationTime = reservationTimeService.getReservationTimeById(reservationTimeId);
+        reservationValidator.validateReservationDateTime(date, reservationTime);
+
+        if (isWaiting) {
+            Reservation waiting = Reservation.createWaiting(member, theme, date, reservationTime);
+            return new ReservationResponse(reservationRepository.save(waiting));
+        }
+        Reservation reserved = Reservation.createReserved(member, theme, date, reservationTime);
+        return new ReservationResponse(reservationRepository.save(reserved));
+    }
```
`createAndSaveReservationBy` 메서드는 하나의 도메인 동작을 수행하는 과정에서

- 서비스로부터 의존 객체를 조회하고,  
- 예약 가능 여부를 검증하며,  
- 예약 혹은 대기 객체를 생성한 뒤,  
- 이를 저장하고 응답을 반환하는 작업까지 수행하고 있습니다.

이처럼 여러 단계의 작업이 하나의 함수에 모두 포함되어 있으면,  
메서드가 어떤 목적을 수행하는지 파악하기 어렵고  
각 단계의 의미가 흐려져 유지보수가 어려워질 수 있습니다.

함수를 분리하는 데 가장 좋은 기준은 **‘목적과 구현을 분리’** 하는 방식입니다.  
코드를 보고 무슨 일을 하는지 파악하는 데 한참이 걸린다면  
그 부분을 함수로 추출한 뒤 **'무슨 일'에 걸맞는 이름을 지어주는 것**이 좋습니다.

코드 블록을 각각의 의미 있는 함수로 추출하면,  
함수의 목적이 더 명확하게 드러나고,  
상위 함수는 각 단계의 흐름을 자연스럽게 서술하는 구조가 되어  
더 이상 신경 쓸 일이 거의 없어집니다.

관련 리팩터링 항목: 『리팩터링 2판』 6.1 _"함수 추출하기 (Extract Function)"_

taek2222 replied:
한 메서드안에 여러 역할을 수행하고 있는 게 느껴저서 아래와 같이 수정했습니다. ☺️

근데, 이래도 머피가 얘기해준 거 처럼 다중 역할을 다 분리하기에는 `byPass` 메서드만 분리되어서 같이 개선 방향도 생각하고 싶습니다. 🤔

**변경 코멘트:**
- [refactor: ReservationCreate 클래스 이름 및 로직 수정](https://github.com/taek2222/spring-roomescape-waiting/commit/000a822b377e33e98b6ff0561e71f58617ff1ba1)

jeyongsong replied:
비타는 객체지향 관점에서 역할과 책임을 고려해주신 것 같네요!  
좋은 접근이라고 생각합니다 :)

하지만, 제가 말씀드렸던 부분은 조금 더 근본적인 클린코드 관점에서의 이야기였습니다 ㅠㅠ ..
"메서드가 어떤 목적을 수행하는지 파악하기 어렵고,  
각 단계의 의미가 흐려져 유지보수가 어려워질 수 있다"는 점에 대해  
한 번 더 고민해보시는 것도 추천드려요!

관련해서 [✨ Clean Code 3장: 함수](https://hyojaedev.tistory.com/40)의 "📌 한 가지만 해라" 파트 와 
[[CleanCode] 함수의 추상화 수준이란?](https://onestone-dev.tistory.com/3)도 
읽어보시면 더 도움이 될 수 있을 것 같아요!

jeyongsong reviewed (commented):
리팩터링 2판 스터디를 진행하며,  
개인적으로 설정한 학습 목표를 실천해보기 위해 작성한 코드 리뷰입니다!
(음... 객관적으로 보이려고 딱딱한 말투를 썼더니, CodeRabbit 느낌 나네 ㅋㅌㅋㅋ)  

> **학습 목표** 
다른 크루의 코드를 보고 내 주관이 아닌 『리팩터링 2판』의 개념을 바탕으로 좋은 리팩터링 방향을 구체적이고 명확하게 제시하는 것

jeyongsong reviewed (commented):
이번 리뷰는 『리팩터링 2판』 스터디를 진행 중인 머피가 아니라,  
비타의 친구 머피로서 작성한 리뷰입니다 ㅋㅌㅋㅋ 🤣
